presto-hive/src/main/java/com/facebook/presto/hive/CacheInvalidationProcedure.java

```diff
@@ -0,0 +1,96 @@
+import com.facebook.presto.spi.procedure.Procedure.Argument;
+import com.google.common.cache.Cache;
+import com.google.common.collect.ImmutableList;
+import org.apache.hadoop.fs.Path;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+
+import java.lang.invoke.MethodHandle;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.common.block.MethodHandleUtil.methodHandle;
+import static com.facebook.presto.common.type.StandardTypes.VARCHAR;
+import static com.facebook.presto.hive.HiveSessionProperties.isUseListDirectoryCache;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_PROCEDURE_ARGUMENT;
+import static java.util.Objects.requireNonNull;
+
+public class CacheInvalidationProcedure
+        implements Provider<Procedure>
+{
+    private static final MethodHandle CACHE_DATA_INVALIDATION = methodHandle(
+            CacheInvalidationProcedure.class,
+            "cacheInvalidation",
+            ConnectorSession.class,
+            String.class);
+
+    private final CachingDirectoryLister cachingDirectoryLister;
+
+    @Inject
+    public CacheInvalidationProcedure(CachingDirectoryLister cachingDirectoryLister)
+    {
+        this.cachingDirectoryLister = requireNonNull(cachingDirectoryLister, "cachingDirectoryLister is null");
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                "system",
+                "invalidate_file_list_cache",
+                ImmutableList.of(
+                        new Argument("directory_path", VARCHAR, false, null)),
+                CACHE_DATA_INVALIDATION.bindTo(this));
+    }
+
+    public void cacheInvalidation(ConnectorSession session, String directoryPath)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
+            doInvalidateDirectoryListCache(session, Optional.ofNullable(directoryPath));
+        }
+    }
+
+    private void doInvalidateDirectoryListCache(ConnectorSession session, Optional<String> directoryPath)
+    {
+        Cache<Path, List<HiveFileInfo>> directoryListCache = cachingDirectoryLister.cache;
+
+        if (isUseListDirectoryCache(session)) {
+            if (directoryPath.isPresent()) {
+                Path path = new Path(directoryPath.get());
+                List<HiveFileInfo> files = directoryListCache.getIfPresent(path);
+                if (files != null) {
+                    directoryListCache.invalidate(path);
+                }
+                else {
+                    throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Given directory path is not cached : " + directoryPath);
+                }
+            }
+            else {
+                cachingDirectoryLister.flushCache();
+            }
+        }
+        else {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Directory list cache is not enabled on catalog");
```

Comment:
nit: Directory list cache is not enabled on this catalog.

Also, imo we should probably have the error messages and procedure naming in sync. We can decide whether we want to call it File List Cache or Directory List Cache.